packages/muelu/src/Misc/MueLu_SegregatedAFactory_def.hpp

```diff
@@ -160,6 +184,129 @@ void SegregatedAFactory<Scalar, LocalOrdinal, GlobalOrdinal, Node>::Build(Level&
   Set(currentLevel, "A", Aout);
 }
 
+template <class Scalar, class LocalOrdinal, class GlobalOrdinal, class Node>
+void SegregatedAFactory<Scalar, LocalOrdinal, GlobalOrdinal, Node>::BuildBasedOnMapPair(Level &currentLevel) const {
+  FactoryMonitor m(*this, "Matrix filtering (segregation, map-pair)", currentLevel);
+
+  RCP<Matrix> Ain = Get<RCP<Matrix>>(currentLevel, "A");
+
+  const ParameterList &pL = GetParameterList();
+  auto comm               = Ain->getRowMap()->getComm();
+
+  // fetch maps from level
+  RCP<const Map> dropMap1 = Teuchos::null;
+  RCP<const Map> dropMap2 = Teuchos::null;
+
+  const std::string dropMap1Name = "dropMap1";
+  const std::string dropMap2Name = "dropMap2";
+
+  if (currentLevel.GetLevelID() == 0) {
+    dropMap1 = currentLevel.Get<RCP<const Map>>(dropMap1Name, NoFactory::get());
+    dropMap2 = currentLevel.Get<RCP<const Map>>(dropMap2Name, NoFactory::get());
+    GetOStream(Statistics0) << "User provided dropMap1 \"" << dropMap1Name << "\": length dimension=" << dropMap1->getGlobalNumElements() << std::endl;
+    GetOStream(Statistics0) << "User provided dropMap2 \"" << dropMap2Name << "\": length dimension=" << dropMap2->getGlobalNumElements() << std::endl;
+  } else {
+    dropMap1 = Get<RCP<const Map>>(currentLevel, dropMap1Name);
+    dropMap2 = Get<RCP<const Map>>(currentLevel, dropMap2Name);
+  }
+
+  TEUCHOS_ASSERT(!dropMap1.is_null());
+  TEUCHOS_ASSERT(!dropMap2.is_null());
+
+  // create new empty Operator
+  Teuchos::RCP<Matrix> Aout = MatrixFactory::Build(Ain->getRowMap(), Ain->getGlobalMaxNumRowEntries());
+
+  Teuchos::RCP<const Map> finalDropMap1 = Teuchos::null;
+  Teuchos::RCP<const Map> finalDropMap2 = Teuchos::null;
+
+  if (pL.get<bool>("Call ReduceAll on dropMap1")) {
+    finalDropMap1 = CreateRedundantMaps(dropMap1);
+  } else {
+    // if reduceAll is not called, we simply work with the local dropMap
+    finalDropMap1 = dropMap1;
+  }
+
+  if (pL.get<bool>("Call ReduceAll on dropMap2")) {
+    finalDropMap2 = CreateRedundantMaps(dropMap2);
+  } else {
+    // if reduceAll is not called, we simply work with the local dropMap
+    finalDropMap2 = dropMap2;
+  }
+
+  // Start copying the matrix row by row and dropping any entries that are contained as a combination of entries of
+  // dropMap1 and dropMap2
+  size_t numLocalMatrixRows = Ain->getLocalNumRows();
+
+  for (size_t row = 0; row < numLocalMatrixRows; row++) {
+    GlobalOrdinal grid = Ain->getRowMap()->getGlobalElement(row);  // global row id
+    bool rowIsInMap1   = finalDropMap1->isNodeGlobalElement(grid);
+    bool rowIsInMap2   = finalDropMap2->isNodeGlobalElement(grid);
+
+    // extract row information from input matrix
+    Teuchos::ArrayView<const LocalOrdinal> indices;
+    Teuchos::ArrayView<const Scalar> vals;
+    Ain->getLocalRowView(row, indices, vals);
+
```

Comment:
Please use the Kokkos Views for new code.

Reply:
I tried to do this but ran into the following problem: 

I believe Xpetra doesn't provide a version of Ain->getLocalRowView that works with Kokkos::Views. As far as I understood, this is only possible in Tpetra exclusive builds. All other implementations expect a Teuchos::ArrayView as input. Our research code still depends on Xpetra, thus the Factory was developed with this in mind.

As far as I can tell all existing use cases in MueLu of getLocalRowView use ArrayViews so my implementation is at least compliant with the existing style. 
Expanding the support of Kokkos in Xpetra goes way beyond of the intention of this pull request. 

If I overlooked or misunderstood something I would be very thankful for any guidance towards the solution. 
Otherwise I suggest to leave it as is for the moment.

Reply:
You could try something like this:
```
auto lclMat = Ain->getLocalMatrixHost();
auto rowView = lclMat.row(rowId);
for (LocalOrdinal jj = 0; jj<rowView.length; ++jj) {
  auto colId = rowView.colidx(jj);
  auto val = rowView.value(jj);
}
```

Reply:
Done